### ISSUE TITLE
Text guided image generation with stable diffusion

### ISSUE THREAD

**Issue by LukeWood** [keras-team/keras-cv#939]
https://twitter.com/_akhaliq/status/1582175757153230849?s=21&t=rCRiXt4-XW41JIyx-jwc-Q

**Comment by LukeWood** [comment]
just opening to triage - added a label for tracking.

**Comment by innat** [comment]
I want to utilize keras-cv to build [this](https://twitter.com/matthieurouif/status/1581969062271520768) product. 😵‍💫  

```bash
segmentation -> stable-diffusion + with-lots-of-tricks
````

**Comment by tanzhenyu** [comment]
We have completed text_to_image, I think the next should be img2img. Closing this request for now

**Comment by innat** [comment]
@tanzhenyu (cc. @LukeWood )
> We have completed text_to_image, 

Actually, text-guided image generation can be categorized one of variant of text-to-image. It's not the same. In text guided image generation, there would be an sample input image too.

![image](https://user-images.githubusercontent.com/17668390/209581626-544634b5-7f4a-4238-84a5-df4c48d6809c.png)

See this for more details. 👉 https://github.com/huggingface/diffusers/issues/1254

**Comment by tanzhenyu** [comment]
> @tanzhenyu (cc. @LukeWood )
> 
> > We have completed text_to_image,
> 
> Actually, text-guided image generation can be categorized one of variant of text-to-image. It's not the same. In text guided image generation, there would be an sample input image too.
> 
> ![image](https://user-images.githubusercontent.com/17668390/209581626-544634b5-7f4a-4238-84a5-df4c48d6809c.png)
> 
> See this for more details. 👉 [huggingface/diffusers#1254](https://github.com/huggingface/diffusers/issues/1254)

Oh I see, so that's the img2img I was referring to then. I am working on it, so re-opening this

**Comment by innat** [comment]
@tanzhenyu (cc. @miguelCalado)
Here is another interesting variant of image-t-image with text guided. Placing it in case you're interested to take it.

Paper: [Null-Text Inversion for Editing Real Images](https://null-text-inversion.github.io/) by Google
Original code: [in PyTorch](https://github.com/google/prompt-to-prompt) (o_O) 
TF 2 code: https://github.com/miguelCalado/prompt-to-prompt-tensorflow (uses keras-cv)

![teaser](https://user-images.githubusercontent.com/17668390/209848471-bf82a2a7-5e05-4486-9be9-e11bb84d5b17.svg)

**Comment by miguelCalado** [comment]
Hi!

Thank you for referring to my implementation of the Prompt-to-Prompt paper @innat. I would be happy to do a PR of the code (after some refactoring) if you guys want 😊
It is a cool method and a useful tool to have in the arsenal when dealing with cross-attention injection, which seems kinda popular these days.

But since the discussion is around text-guided image generation, why don't you start by adding negative prompting? It seems to be useful, especially when dealing with SD 2.x.

**Comment by tanzhenyu** [comment]
> Hi!
> 
> Thank you for referring to my implementation of the Prompt-to-Prompt paper @innat. I would be happy to do a PR of the code (after some refactoring) if you guys want 😊 It is a cool method and a useful tool to have in the arsenal when dealing with cross-attention injection, which seems kinda popular these days.
> 
> But since the discussion is around text-guided image generation, why don't you start by adding negative prompting? It seems to be useful, especially when dealing with SD 2.x.

That would be great! Do you want to start with negative prompting, or should I? (I have been busy with the 0.4 release so this might take me 2 weeks)

**Comment by miguelCalado** [comment]
Sure! It would be my pleasure!

I opened an issue #1206 for further discussion.

**Comment by innat** [comment]
@miguelCalado Congrats for the 1st place of [keras community price.](https://discuss.tensorflow.org/t/announcing-the-keras-community-prize-first-edition/13148/29?u=innat) 👍

**Comment by tanzhenyu** [comment]
> keras community price.

Congrats @miguelCalado !! We really appreciate your work. If you have other ideas to improve our existing offering, please go ahead!

**Comment by miguelCalado** [comment]
Thanks everyone for the wishes! I'm still in shock and this whole thing hasn't really settled down :sweat_smile:

Yes, I'm looking forward to contributing some more. It would be cool to see implemented in KerasCV multi-prompting - adding weights to parts of the prompts, different schedulers, more versions of stable diffusion (e.g. 1.5), other research works (e.g. [image variation](https://huggingface.co/spaces/lambdalabs/stable-diffusion-image-variations) or [Imagic](https://arxiv.org/abs/2210.09276)), etc. There is a lot of room for contributions :grin:

But one PR at a time! The Prompt-to-prompt will take me a bit as I'm still giving the final touches (adding support for multiple batches and other small things :slightly_smiling_face:).

Thanks everyone!

**Comment by miguelCalado** [comment]
![teaser](https://user-images.githubusercontent.com/37347167/216439836-9a7b0584-dba4-430c-952a-9ff4e313c79f.jpg)

Continuing the thread of text-guided image generation, this work also looks interesting: "it refines the cross-attention units to attend to all subject tokens in the text prompt and strengthen - or excite - their activations, encouraging the model to generate all subjects described in the text prompt.". The [video](https://www.youtube.com/watch?v=9EWs2IX4cus&ab_channel=Attend-and-Excite) summarizes it pretty well

It appears simple to implement (no training/finetuning), and it could work as a parameter on the `text_to_image` method (e.g. `apply_excite_tokens=[2,5]`). Might have a go after being done with the Prompt-to-prompt PR!

**Comment by Elvenson** [comment]
Hi guys, I wonder if this issue is still active. But I found a helpful reference for the img2img implementation [here](https://github.com/divamgupta/stable-diffusion-tensorflow/blob/master/img2img.py). I also integrated this logic into my [branch](https://github.com/Elvenson/keras-cv/pull/3/files) and tested it. I hope it can help you guys.

**Comment by jbischof** [comment]
Thank you @Elvenson! This had been deprioritized but we'll take a look at your code.

**Comment by sachinprasadhs** [comment]
Thanks for reporting the issue! We have consolidated the development of KerasCV into the new **KerasHub** package, which supports image, text, and multi-modal models. Please read [the announcement](https://github.com/keras-team/keras-hub/issues/1831). `KerasHub` will support all the core functionality of `KerasCV`.

`KerasHub` can be installed with `!pip install -U keras-hub`. Documentation and guides are available at [keras.io/keras_hub](https://keras.io/keras_hub/).

With our focus shifted to KerasHub, we are not planning any further development or releases in KerasCV. If you encounter a KerasCV feature that is missing from KerasHub, or would like to propose an addition to the library, please file [an issue with KerasHub](https://github.com/keras-team/keras-hub/issues).
